### PR TITLE
Build Teleport Connect on darwin/amd64

### PR DESCRIPTION
**Pull request by zmb3**
This commit updates drone to build Teleport Connect by:

- cloning `gravitational/webapps` as a sibling directory to gravitational/teleport
- checkout out the right version of webapps by running a simple Go program (this step is only necessary until we move `webapps` into the teleport repo)
- Running the Teleport Connect build and copying artifacts

Code signing should run on tag builds automatically as part the electron build, assuming the Apple Account credentials are properly loaded into the keychain.

Notarization will also happen automatically if both `$APPLE_USERNAME` and `$APPLE_PASSWORD` are set.

In order to make the above happen, this patch also includes:

- Installing and removing a per-build Node instance in the toolchain directory on Darwin
- Moving the toolchain temporary directory out of `~/` and into `/tmp`. 

Drone usually sets `$HOME` to a temporary directory for each build, but unfortunately we need it to point to the actual `build` user's home directory in order for the notarisation tooling to find the right keychain. Having `$HOME` point to a long-lived directory risks both pollution from build detritus and builds stomping on one another. In an in an attempt to isolate the builds from each other and protect `~build` as best we can, as much of the build state as possible (including ephemeral toolchains) has been moved under `/tmp`.

TODO:

- [ ] make the corresponding changes to e/.drone.yml

### CODE REVIEW

File: build.assets/tooling/cmd/get-webapps-version/main.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	// tag exists on gravitational/webapps
+	if tag != "" {
+		return tag
+	}
+
+	// if this build is on one of the teleport release branches,
+	// map to the equivalent release branch in webapps
+	if strings.HasPrefix(targetBranch, "branch/") {
+		return "teleport-" + strings.TrimPrefix(targetBranch, "branch/")
+	}
```
Comment (wadells) [inline build.assets/tooling/cmd/get-webapps-version/main.go:30]: I don't love that this leads to a floating dependency (latest commit on the branch), which will then result in nondeterminism during any untagged builds.  That said, I understand this is an interim solution, and updating the equivalent of a lockfile is tedious.

**Review by wadells — COMMENTED** [review]
Initial thoughts: Could we have teleport connect build as a separate pipelines?  I think that may be logically clearer, even if it results in duplication of some build steps.  The extra machine time isn't expensive, and being able to automatically separate/diagnose that any failures are connect related would be valuable.

**Comment by zmb3** [comment]
> Initial thoughts: Could we have teleport connect build as a separate pipelines?  I think that may be logically clearer, even if it results in duplication of some build steps.  The extra machine time isn't expensive, and being able to automatically separate/diagnose that any failures are connect related would be valuable.

Perhaps, but Teleport Connect "vendors" in a copy of the tsh binary, and I thought it would be best if the binary was exactly the same one as the one in the teleport release.

File: build.assets/tooling/cmd/get-webapps-version/main.go (hunk 1 above)
Comment (zmb3) [reply build.assets/tooling/cmd/get-webapps-version/main.go:30]: Yeah, this isn't great, but it at least ensures that connect continues to build successfully, so it's better than nothing which is the state we're in today.

File: build.assets/Dockerfile-teleterm [hunk 2]
```diff
@@ -15,7 +15,7 @@ RUN BIN="/usr/local/bin" && \
     chmod +x "${BIN}/${BINARY_NAME}"
 
 # Install node
-ARG NODE_VERSION=v15.14.0
+ARG NODE_VERSION=v16.14.2
```
Comment (ravicious) [inline build.assets/Dockerfile-teleterm:18]: This change is irrelevant to the rest of the changes, right? Drone won't be building this inside Docker?

Which makes me wonder: how are we going to ensure that Drone has the correct Node version?

File: dronegen/mac.go [hunk 3]
```diff
@@ -270,18 +295,45 @@ func darwinTagBuildCommands() []string {
...
 		`cp teleport*.tar.gz $WORKSPACE_DIR/go/artifacts`,
 		`cp e/teleport-ent*.tar.gz $WORKSPACE_DIR/go/artifacts`,
-		// generate checksums (for mac)
+	}
+
+	// copy Teleport Connect artifacts
+	if b.hasTeleportConnect() {
+		commands = append(commands,
+			`cd $WORKSPACE_DIR/go/src/github.com/gravitational/webapps/packages/teleterm/build/release`,
+			`cp teleterm*.dmg $WORKSPACE_DIR/go/artifacts`,
```
Comment (ravicious) [inline dronegen/mac.go:322]: The artifacts don't start with `teleterm`, they start with whatevers in `packages/teleterm/package.json` under the `productName` key. Right now it's `Teleport Connect Preview`.

In theory we could read that with something like jq, which is available in teleterm dockerfile, but I'm not sure about Drone. We'll going to change it only one more time though with the GA release, so idk if it even makes sense to do that.

File: dronegen/mac.go [hunk 4]
```diff
@@ -270,18 +295,45 @@ func darwinTagBuildCommands() []string {
...
+		)
+	}
+
+	// generate checksums
+	commands = append(commands,
 		`cd $WORKSPACE_DIR/go/artifacts && for FILE in teleport*.tar.gz; do shasum -a 256 $FILE > $FILE.sha256; done && ls -l`,
+	)
+	if b.hasTeleportConnect() {
+		commands = append(commands,
+			`cd $WORKSPACE_DIR/go/artifacts && for FILE in teleterm*.dmg; do shasum -a 256 $FILE > $FILE.sha256; done && ls -l`,
```
Comment (ravicious) [inline dronegen/mac.go:332]: The artifact glob needs changing here as well.

File: build.assets/Dockerfile-teleterm (hunk 2 above)
Comment (zmb3) [reply build.assets/Dockerfile-teleterm:18]: Correct, but our README says node16 so I figured I would change this while I'm at it.

Drone will build the linux builds in docker when we get there.

For the mac builders, the node version is set in a separate PR: https://github.com/gravitational/ops/pull/355

File: dronegen/mac.go (hunk 3 above)
Comment (zmb3) [reply dronegen/mac.go]: Thanks. I agree since it's not going to change often let's just hard code it.

**Review by ravicious — APPROVED** [review]

File: build.assets/tooling/cmd/get-webapps-version/main.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+func main() {
+	fmt.Println(webappsVersion(
```
Comment (r0mant) [inline build.assets/tooling/cmd/get-webapps-version/main.go:26]: Should this strip whitespaces (or just fmt.Print instead) so the caller doesn't have to deal with newlines?

File: build.assets/tooling/cmd/get-webapps-version/main.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func webappsVersion(tag, targetBranch string) string {
+	// if this build was triggered from a tag on the
+	// gravitational/teleport repo, assume that same
+	// tag exists on gravitational/webapps
+	if tag != "" {
+		return tag
+	}
+
+	// if this build is on one of the teleport release branches,
+	// map to the equivalent release branch in webapps
```
Comment (r0mant) [inline build.assets/tooling/cmd/get-webapps-version/main.go:41]: Nit: I'd give a specific example in the comment e.g.

```
For example, branch/v9 in teleport repo maps onto teleport-v9 in web apps.
```

Would make it a bit easier to understand without parsing the code.

**Review by r0mant — APPROVED** [review]

**Comment by tcsc** [comment]
> > Initial thoughts: Could we have teleport connect build as a separate pipelines?  I think that may be logically clearer, even if it results in duplication of some build steps.  The extra machine time isn't expensive, and being able to automatically separate/diagnose that any failures are connect related would be valuable.
> 
> Perhaps, but Teleport Connect "vendors" in a copy of the tsh binary, and I thought it would be best if the binary was exactly the same one as the one in the teleport release.

I realise I'm probably late to the party here, but you _could_ split the current build-and-release pipeline up into several smaller chunks that have more discrete jobs. I'm imagining something like this, where "Build Teleport", "Build Teleport Connect" and "Bundle release" are all separate pipelines with the appropriate inter-dependencies set up.

```
     teleport       Staging
   ┌───tctl──────►   Area   ───────────┐
   │   tsh           │  ▲              │
   │            ┌────┘  └────┐        all
   │            │            │      binaries
   │            │         teleport     │
               tsh         connect     ▼
 Build          │            │       Bundle
Teleport        │   Build    │      Release ─────tarball──────►
                └► Teleport ─┘
                   Connect
```

Although this may be getting a bit overblown for a temporary solution.

File: dronegen/common.go [hunk 7]
```diff
@@ -162,6 +162,10 @@ func (b *buildType) Description(packageType string, extraQualifications ...strin
 	return result
 }
 
+func (b *buildType) hasTeleportConnect() bool {
+	return b.os == "darwin" && b.arch == "amd64"
```
Comment (tcsc) [inline dronegen/common.go:166]: Is teleport connect mac only?

**Review by tcsc — APPROVED** [review]

File: dronegen/common.go (hunk 7 above)
Comment (zmb3) [reply dronegen/common.go:166]: For now, yes it is darwin/amd64 only.

**Comment by ravicious** [comment]
At some point, we want to do two things:

* Rename `packages/teleterm` in the `webapps` repo to something else.
   * `teleterm` is a legacy term while Teleport Connect was still called Teleport Terminal.
   * Rather unlikely to happen during the next month.
* Change the name of the packaged app from `Teleport Connect Preview` to `Teleport Connect`.
   * This will happen for the GA release of Connect which is supposed to go out with the v10 release.
   * Can be accessed under the `productName` key in `packages/teleterm/package.json`, so it depends on the previous item.

Should we future-proof our build scripts in this PR so that in the future we'll only have to change the code in the webapps repo without having to sync this with a change to the build servers?

We could add a custom key to the main `package.json` in the webapps repo so that it points at the appropriate directory. We could also add npm scripts that extract that information to avoid having to use jq as a dependency. This means that in the build scripts we could extract that by grabbing output from commands, for example by running `yarn teleport-connect-dir` and `yarn teleport-connect-product-name`.

**Comment by tcsc** [comment]
> Should we future-proof our build scripts in this PR so that in the future we'll only have to change the code in the webapps repo without having to sync this with a change to the build servers?

Makes sense. Do we know if we produce any other `dmg`s that we _don't_ want included in the release? If not, I can just simplify the packaging script to grab `*.dmg` and whatever is in the release dir will go into the package.

File: build.assets/Dockerfile-teleterm (hunk 2 above)
Comment (tcsc) [reply build.assets/Dockerfile-teleterm:18]: It's now set in the `build.assets` makefile, with all the other tool versions

**Comment by ravicious** [comment]
> Do we know if we produce any other `dmg`s that we _don't_ want included in the release?

This is the only `dmg` that we produce.

I'll prepare a PR tomorrow in the webapps repo that adds a top-level `teleport-connect-dir` script so that in the build scripts we will be able to execute `yarn -s teleport-connect-dir` to get the directory.

```ruby
irb(main):001:0> `yarn -s teleport-connect-dir`
=> "packages/teleterm\n"
```

**Review by ravicious — APPROVED** [review]

File: dronegen/mac.go [hunk 8]
```diff
@@ -251,37 +312,83 @@ func cleanUpExecStorageStep(path string) step {
...
 		`cp teleport*.tar.gz $WORKSPACE_DIR/go/artifacts`,
 		`cp e/teleport-ent*.tar.gz $WORKSPACE_DIR/go/artifacts`,
-		// generate checksums (for mac)
+	}
+
+	// copy Teleport Connect artifacts
+	if b.hasTeleportConnect() {
+		commands = append(commands,
+			`cd $WORKSPACE_DIR/go/src/github.com/gravitational/webapps/packages/teleterm/build/release`,
+			`cp *.dmg $WORKSPACE_DIR/go/artifacts`,
```
[the remark this answers is not in the document]
Comment (ravicious) [reply dronegen/mac.go:377]: Ahh nvm, I just saw #12622, I think that confirms it.

File: build.assets/Dockerfile-teleterm (hunk 2 above)
Comment (tcsc) [reply build.assets/Dockerfile-teleterm:18]: ...and passed in from the `build.assets` makefile

**Comment by ravicious** [comment]
Now that I think about it, the Connect-related steps probably shouldn't run on v8 and earlier branches because they don't include code for Connect. Do we need to add a check for that as well?

**Review by zmb3 — COMMENTED** [review]
Can't approve my own PR, but looks good.  Thanks Trent for picking this up!

**Comment by tcsc** [comment]
> Now that I think about it, the Connect-related steps probably shouldn't run on v8 and earlier branches because they don't include code for Connect. Do we need to add a check for that as well?

If we don't backport the Drone changes, the old branches shouldn't get the build steps.

File: dronegen/mac.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply dronegen/mac.go:381]: Yep - but it will take some work in Houston and the website code to make it work.

**Pull request merged**